components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -765,6 +773,169 @@ IJsonSerializer jsonSerializer
+            await _lockConfig.WaitAsync(ct).ConfigureAwait(false);
+            try {
+                // ToDo: Uncomment ValidateRequest() once our test requests pass validation.
+                // This will throw SerializerException if values of request fields are not conformant.
+                //ValidateRequest(request);
+
+                // First let's check that endpoints that we are asked to remove exist.
+                var dataSetsToRemove = request.Where(e => e.OpcNodes is null || e.OpcNodes.Count == 0).ToList();
+                var dataSetsToUpdate = request.Where(e => !(e.OpcNodes is null) && e.OpcNodes.Count > 0).ToList();
+
+                foreach (var dataSetToRemove in dataSetsToRemove) {
+                    var foundDataSet = false;
+                    foreach (var entry in _publishedNodesEntries) {
+                        foundDataSet = foundDataSet || IsSameDataSet(entry, dataSetToRemove);
+                    }
+                    if (!foundDataSet) {
+                        throw new MethodCallStatusException((int)HttpStatusCode.NotFound,
+                            $"Endpoint not found: {dataSetToRemove.EndpointUrl}");
+                    }
+                }
+
+                var existingGroups = new List<PublishedNodesEntryModel>();
+                var requestDataSetsFound = Enumerable.Repeat(false, request.Count).ToList(); ;
+
+                foreach (var entry in _publishedNodesEntries) {
+                    var groupFound = false;
+
+                    for (var k = 0; k < request.Count; ++k) {
+                        var dataSetToUpdate = request[k];
+                        if (entry.HasSameGroup(dataSetToUpdate)) {
+                            groupFound = true;
+
+                            // We may have several entries with the same DataSetGroup definition,
+                            // so we will update nodes only if the whole DataSet definition matches.
+                            if (IsSameDataSet(entry, dataSetToUpdate)) {
+                                if (dataSetToUpdate.OpcNodes is null || dataSetToUpdate.OpcNodes.Count == 0 || requestDataSetsFound[k]) {
+                                    // In this case existing OpcNodes entries should be cleaned up.
+                                    entry.OpcNodes.Clear();
+                                }
+                                else {
+                                    // We will add OpcNodes to the first matching entry
+                                    // and the rest will be cleaned up.
+                                    entry.OpcNodes = dataSetToUpdate.OpcNodes;
+                                }
+
+                                // We do not need to look for another matching data set in request.
+                                requestDataSetsFound[k] = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    if (groupFound) {
```

Comment:
>groupFound [](http://example.com/codeflow?start=24&length=10)

isn't this the other way around? at least the description implies this.

Reply:
yeah, it shows if a matching data set group was found (in `request`) for a specific entry in `_publishedNodesEntries`.

Reply:
now I got it